File: TestSandbox.php
<?php

getStatus();
//$newId = createProfile("bqj24@tester.com", "+447917654422"); // email and phone number need to be unique
//echo "\nnewId: $newId\n";
//getProfile($newId);
//getAllProfiles();
//deleteProfile($newId);
//getAllProfiles();


// Get the status of the Customate sandbox (GET https://sandbox-api.gocustomate.com/v1/status)
function getStatus() {
    $date = (new Datetime()) -> format('Y-m-d\TH:i:s\Z');
    $uuid = getGUID();

    $string_to_sign = utf8_encode("GET\n/v1/status\n\npaymentservice-contenthash:\npaymentservice-date:$date\npaymentservice-nonce:$uuid");
    echo "****************************************\n\nRequest\n-------\n$string_to_sign\n";

    $access_token = createToken($string_to_sign, getApiSecret());
    echo "\nAccess Token: $access_token\n";
    $signature = getApiKey() . ":" . $access_token;

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, getApiUrl()."/status");
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($ch, CURLOPT_HEADER, FALSE);
    curl_setopt($ch, CURLOPT_HTTPHEADER, [
        "PaymentService-Date:$date",
        "PaymentService-Nonce:$uuid",
        "Authorization: Signature $signature"
    ]);

    $response = curl_exec($ch);
    echo "\nResponse\n--------\n";
    var_dump($response);
    curl_close($ch);
}


// Get profiles (https://sandbox-api.gocustomate.com/v1/profiles?page[number]=1&page[size]=99)
function getAllProfiles() {
    $date = (new Datetime()) -> format('Y-m-d\TH:i:s\Z');
    $uuid = getGUID();
    $string_to_sign = utf8_encode("GET\n/v1/profiles\n\npaymentservice-contenthash:\npaymentservice-date:$date\npaymentservice-nonce:$uuid");

    echo "\n****************************************\n\nRequest\n-------\n$string_to_sign\n";
    $access_token = createToken($string_to_sign, getApiSecret());
    echo "\nAccess Token: $access_token\n";
    $signature = getApiKey() . ":" . $access_token;

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, getApiUrl()."/profiles");
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($ch, CURLOPT_HEADER, FALSE);
    curl_setopt($ch, CURLOPT_HTTPHEADER, [
        "PaymentService-Date:$date",
        "PaymentService-Nonce:$uuid",
        "Authorization: Signature $signature"
    ]);

    $response = curl_exec($ch);
    echo "\nResponse\n--------\n";
    var_dump($response);
    curl_close($ch);
}


// Create a profile - email and phone number need to be unique (POST https://sandbox-api.gocustomate.com/v1/profiles)
function createProfile($email, $phone_number) {
    $date = (new Datetime()) -> format('Y-m-d\TH:i:s\Z');
    $uuid = getGUID();

    $body = [
      "type" => "personal",
      "first_name" => "Bob",
      "middle_name" => "Quentin",
      "last_name" => "Jones",
      "email" => $email,
      "phone_number" => $phone_number,
      "birth_date" => "1980-01-01",
      "title" => "mr",
      "gender" => "male",
      "address" => [
        "line_1" => "117 Boroughbridge Road",
        "line_2" => "Near Edgbaston",
        "line_3" => "Edgbaston",
        "city" => "Birmingham",
        "locality" => "Birmingham",
        "postcode" => "B5 7DS",
        "country" => "GB",
      ],
      "metadata" => [
        "sample_internal_id" => "505f7b6e-d0e9-46da-a2f9-e2a737a29d19"
      ]
    ];

    $requestBody = json_encode($body);
    $contentHashSource = utf8_encode($requestBody);
    $string_to_sign = utf8_encode("POST\n/v1/profiles\napplication/json\npaymentservice-contenthash:\npaymentservice-date:$date\npaymentservice-nonce:$uuid");

    echo "****************************************\n\nRequest\n-------\n$string_to_sign\n";
    $access_token = createToken($string_to_sign, getApiSecret());
    $signature = getApiKey() . ":" . $access_token;

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, getApiUrl()."/profiles");
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($ch, CURLOPT_HEADER, FALSE);
    curl_setopt($ch, CURLOPT_POST, TRUE);
    curl_setopt($ch, CURLOPT_HTTPHEADER, [
        "Content-Type: application/json",
        "PaymentService-Date:$date",
        "PaymentService-Nonce:$uuid",
        "Authorization: Signature $signature"
    ]);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $contentHashSource);

    $response = curl_exec($ch); // The response returns: {"id":"<36 character id>", ...
    $newId = substr($response, 7, 36);
    echo "\nResponse\n--------\n";
    var_dump($response);
    curl_close($ch);
    return $newId;
}


// Get the profile (https://sandbox-api.gocustomate.com/v1/profiles/<profile_id>)
function getProfile($id) {
    $date = (new Datetime()) -> format('Y-m-d\TH:i:s\Z');
    $uuid = getGUID();
    $string_to_sign = utf8_encode("GET\n/v1/profiles/$id\n\npaymentservice-contenthash:\npaymentservice-date:$date\npaymentservice-nonce:$uuid");

    echo "\n****************************************\n\nRequest\n-------\n$string_to_sign\n";
    $access_token = createToken($string_to_sign, getApiSecret());
    echo "\nAccess Token: $access_token\n";
    $signature = getApiKey() . ":" . $access_token;

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, getApiUrl()."/profiles/$id");
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($ch, CURLOPT_HEADER, FALSE);
    curl_setopt($ch, CURLOPT_HTTPHEADER, [
        "PaymentService-Date:$date",
        "PaymentService-Nonce:$uuid",
        "Authorization: Signature $signature"
    ]);

    $response = curl_exec($ch);
    echo "\nResponse\n--------\n";
    var_dump($response);
    curl_close($ch);
}


// Delete the profile (DELETE https://sandbox-api.gocustomate.com/v1/profiles/<profile_id>)
function deleteProfile($id) {
    $date = (new Datetime()) -> format('Y-m-d\TH:i:s\Z');
    $uuid = getGUID();
    $string_to_sign = utf8_encode("DELETE\n/v1/profiles/$id\n\npaymentservice-contenthash:\npaymentservice-date:$date\npaymentservice-nonce:$uuid");

    echo "\n****************************************\n\nRequest\n-------\n$string_to_sign\n";
    $access_token = createToken($string_to_sign, getApiSecret());
    echo "\nAccess Token: $access_token\n";
    $signature = getApiKey() . ":" . $access_token;

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, getApiUrl()."/profiles/$id");
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($ch, CURLOPT_HEADER, FALSE);
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "DELETE");
    curl_setopt($ch, CURLOPT_HTTPHEADER, [
        "PaymentService-Date:$date",
        "PaymentService-Nonce:$uuid",
        "Authorization: Signature $signature"
    ]);

    $response = curl_exec($ch);
    echo "\nResponse\n--------\n";
    var_dump($response);
    curl_close($ch);
}


// Generate a Guid (from: https://stackoverflow.com/questions/21671179/how-to-generate-a-new-guid)
function getGUID() {
    if (function_exists('com_create_guid')){
        return com_create_guid();
    } else {
        mt_srand((double) microtime() * 10000); // Optional for php 4.2.0 and up.
        $charid = strtoupper(md5(uniqid(rand(), true)));
        $hyphen = chr(45); // "-"
        $uuid = ""
            .substr($charid,  0,  8).$hyphen
            .substr($charid,  8,  4).$hyphen
            .substr($charid, 12,  4).$hyphen
            .substr($charid, 16,  4).$hyphen
            .substr($charid, 20, 12);
        return $uuid;
    }
}


function getApiUrl() {
    return 'https://sandbox-api.gocustomate.com/v1';
}


function getApiKey() {
    return '<YOUR API KEY>';
}


function getApiSecret() {
    return '<YOUR API SECRET>';
}


// Encrypts a string using the secret
function createToken($message, $secret) {
    $hmac_hash = hash_hmac("sha256", $message, $secret);
    $access_token = utf8_decode(base64_encode(pack('H*', $hmac_hash)));
    return $access_token;
}
?>

File: TestSandbox.cs
using System;
using System.IO;
using System.Net;
using System.Security.Cryptography;
using System.Text;

public class TestSandbox {

    private static String apiUrl = "https://sandbox-api.gocustomate.com/v1";
    private static String apiKey = "01331fd1-d4d4-4ca4-a457-7635d7ce4a3e";
    private static String apiSecret = "c1e1aaf3ca4b44a7cc634939e74c260893e9d225961ce45098";

    public static void Main() {
        GetAPIStatus();
        //GetAllProfiles();
        //String newId = CreateProfile("bqj18@tester.com", "+447917654403"); // email and phone need to be unique
        //Console.WriteLine("\nnewId: " + newId + "\n");
        //GetProfile(newId);
        //UpdateProfile(newId, "bqj19@tester.com", "+447917654404");
        //DeleteProfile(newId);
    }


    // Encrypts a string using the secret
    private static string CreateToken(string message, string secret)
    {
        secret = secret ?? "";
        var encoding = new System.Text.ASCIIEncoding();
        byte[] keyByte = encoding.GetBytes(secret);
        byte[] messageBytes = encoding.GetBytes(message);
        using (var hmacsha256 = new HMACSHA256(keyByte))
        {
            byte[] hashmessage = hmacsha256.ComputeHash(messageBytes);
            String base64String = Convert.ToBase64String(hashmessage);
            return base64String;
        }
    }


    // GET the status of the Customate sandbox (https://sandbox-api.gocustomate.com/v1/status)
    private static void GetAPIStatus()
    {
        String datetime = DateTime.UtcNow.ToString("s") + "Z";
        Guid guid = Guid.NewGuid();

        String message = "GET\n/v1/status\n\npaymentservice-contenthash:\npaymentservice-date:" + datetime + "\npaymentservice-nonce:" + guid;
        String token = CreateToken(message, apiSecret);
        System.Console.WriteLine("****************************************\n\nRequest\n-------\n" + message);
        System.Console.WriteLine("\nAccess Token:: " + token);

        string url = apiUrl + "/status";
        string output = string.Empty;

        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.Headers["PaymentService-Date"] = datetime;
        request.Headers["PaymentService-Nonce"] = guid.ToString();
        request.Headers["Authorization"] = "Signature " + apiKey + ":" + token;

        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        using (Stream stream = response.GetResponseStream())
        using (StreamReader reader = new StreamReader(stream))
        {
            output = reader.ReadToEnd();
        }
        Console.WriteLine("\nResponse\n--------\n" + output + "\n");
    }


    // Get all profiles (GET https://sandbox-api.gocustomate.com/v1/profiles)
    private static void GetAllProfiles() {
        String datetime = DateTime.UtcNow.ToString("s") + "Z";
        Guid guid = Guid.NewGuid();

        String message = "GET\n/v1/profiles\n\npaymentservice-contenthash:\npaymentservice-date:" + datetime + "\npaymentservice-nonce:" + guid;
        String token = CreateToken(message, apiSecret);
        System.Console.WriteLine("****************************************\n\nRequest\n-------\n" + message);
        System.Console.WriteLine("\nAccess Token:: " + token);

        string url = apiUrl + "/profiles";
        string output = string.Empty;

        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.Headers["PaymentService-Date"] = datetime;
        request.Headers["PaymentService-Nonce"] = guid.ToString();
        request.Headers["Authorization"] = "Signature " + apiKey + ":" + token;

        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        using (Stream stream = response.GetResponseStream())
        using (StreamReader reader = new StreamReader(stream))
        {
            output = reader.ReadToEnd();
            Console.WriteLine("\nResponse\n--------\n" + output);
        }
    }


    // Creates a profile - the email and phone number need to be unique (POST https://sandbox-api.gocustomate.com/v1/profiles)
    private static String CreateProfile(String email, String phoneNumber) {
        String body =
            "{" +
                "\"type\" : \"personal\", " +
                "\"first_name\" : \"Bobby\", " +
                "\"middle_name\" : \"Quentino\", " +
                "\"last_name\" : \"Jonesy\", " +
                "\"email\" : \"" + email + "\", " +
                "\"phone_number\" : \"" + phoneNumber + "\", " +
                "\"birth_date\" : \"1980-01-02\", " +
                "\"title\" : \"mrs\", " +
                "\"gender\" : \"female\", " +
                "\"address\" : {" +
                    "\"line_1\" : \"118 Boroughbridge Road\", " +
                    "\"line_2\" : \"Near Bourneville\", " +
                    "\"line_3\" : \"Birmingham\", " +
                    "\"city\" : \"Birmingham\", " +
                    "\"locality\" : \"Birmingham\", " +
                    "\"postcode\" : \"B5 7DT\", " +
                    "\"country\" : \"GB\"" +
                "}, " +
                "\"metadata\" : {" +
                    "\"sample_internal_id\" : \"505f7b6e-d0e9-46da-a2f9-e2a737a29d19\"" +
                "}" +
            "}";

        byte[] bytes = Encoding.Default.GetBytes(body);
        String encodedBody = Encoding.UTF8.GetString(bytes);

        String datetime = DateTime.UtcNow.ToString("s") + "Z";
        Guid guid = Guid.NewGuid();

        String message = "POST\n/v1/profiles\napplication/json\npaymentservice-contenthash:\npaymentservice-date:" + datetime + "\npaymentservice-nonce:" + guid;
        String token = CreateToken(message, apiSecret);
        System.Console.WriteLine("****************************************\n\nRequest\n-------\n" + message);
        System.Console.WriteLine("\nAccess Token:: " + token);

        String url = apiUrl + "/profiles";
        String output = string.Empty;

        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.Method = "POST";
        request.ContentType = "application/json";
        request.Headers["PaymentService-Date"] = datetime;
        request.Headers["PaymentService-Nonce"] = guid.ToString();
        request.Headers["Authorization"] = "Signature " + apiKey + ":" + token;

        using (var streamWriter = new StreamWriter(request.GetRequestStream()))
        {
            streamWriter.Write(encodedBody);
        }

        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        using (Stream stream = response.GetResponseStream())
        using (StreamReader reader = new StreamReader(stream))
        {
            output = reader.ReadToEnd();
        }
        Console.WriteLine("\nResponse\n--------\n" + output);

        // The output returns: {"id":"<36 character id>", ...
        String newId = output.Substring(7, 36);
        return newId;
    }


    // Get the profile (GET https://sandbox-api.gocustomate.com/v1/profiles/<profile_id>)
    private static void GetProfile(String id) {
        String datetime = DateTime.UtcNow.ToString("s") + "Z";
        Guid guid = Guid.NewGuid();

        String message = "GET\n/v1/profiles/" + id + "\n\npaymentservice-contenthash:\npaymentservice-date:" + datetime + "\npaymentservice-nonce:" + guid;
        String token = CreateToken(message, apiSecret);
        System.Console.WriteLine("****************************************\n\nRequest\n-------\n" + message);
        System.Console.WriteLine("\nAccess Token:: " + token);

        string url = apiUrl + "/profiles/" + id;
        string output = string.Empty;

        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.Headers["PaymentService-Date"] = datetime;
        request.Headers["PaymentService-Nonce"] = guid.ToString();
        request.Headers["Authorization"] = "Signature " + apiKey + ":" + token;

        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        using (Stream stream = response.GetResponseStream())
        using (StreamReader reader = new StreamReader(stream))
        {
            output = reader.ReadToEnd();
        }
        Console.WriteLine("\nResponse\n--------\n" + output);
    }


    // Update the profile (PUT https://sandbox-api.gocustomate.com/v1/profiles/<profile_id>)
    private static void UpdateProfile(String id, String email, String phoneNumber) {
        String body =
            "{" +
                "\"id\" : \"" + id + "\", " +
                "\"creation_datetime\" : \"2020-10-19T11:02:29.258312\", " +
                "\"type\" : \"personal\", " +
                "\"first_name\" : \"Bobby\", " +
                "\"middle_name\" : \"Quentino\", " +
                "\"last_name\" : \"Jonesy\", " +
                "\"email\" : \"" + email + "\", " +
                "\"phone_number\" : \"" + phoneNumber + "\", " +
                "\"birth_date\" : \"1980-01-02\", " +
                "\"title\" : \"mrs\", " +
                "\"gender\" : \"female\", " +
                "\"address\" : {" +
                    "\"line_1\" : \"119 Cricket Road\", " +
                    "\"line_2\" : \"Near Edgbaston\", " +
                    "\"line_3\" : \"Edgbaston\", " +
                    "\"city\" : \"Birmingham\", " +
                    "\"locality\" : \"Birmingham\", " +
                    "\"postcode\" : \"B5 7DU\", " +
                    "\"country\" : \"GB\"" +
                "}, " +
                "\"metadata\" : {" +
                    "\"sample_internal_id\" : \"505f7b6e-d0e9-46da-a2f9-e2a737a29d19\"" +
                "}" +
            "}";

        byte[] bytes = Encoding.Default.GetBytes(body);
        String encodedBody = Encoding.UTF8.GetString(bytes);

        String datetime = DateTime.UtcNow.ToString("s") + "Z";
        Guid guid = Guid.NewGuid();

        String message = "PUT\n/v1/profiles/" + id + "\napplication/json\npaymentservice-contenthash:\npaymentservice-date:" + datetime + "\npaymentservice-nonce:" + guid;
        String token = CreateToken(message, apiSecret);
        System.Console.WriteLine("****************************************\n\nRequest\n-------\n" + message);
        System.Console.WriteLine("\nAccess Token:: " + token);

        String url = apiUrl + "/profiles/" + id;
        String output = string.Empty;

        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.Method = "PUT";
        request.ContentType = "application/json";
        request.Headers["PaymentService-Date"] = datetime;
        request.Headers["PaymentService-Nonce"] = guid.ToString();
        request.Headers["Authorization"] = "Signature " + apiKey + ":" + token;

        using (var streamWriter = new StreamWriter(request.GetRequestStream()))
        {
            streamWriter.Write(encodedBody);
        }

        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        using (Stream stream = response.GetResponseStream())
        using (StreamReader reader = new StreamReader(stream))
        {
            output = reader.ReadToEnd();
        }
        Console.WriteLine("\nResponse\n--------\n" + output);
    }


    // Delete the profile (DELETE https://sandbox-api.gocustomate.com/v1/profiles/<profile_id>)
    private static void DeleteProfile(String id) {
        String datetime = DateTime.UtcNow.ToString("s") + "Z";
        Guid guid = Guid.NewGuid();

        String message = "DELETE\n/v1/profiles/" + id + "\n\npaymentservice-contenthash:\npaymentservice-date:" + datetime + "\npaymentservice-nonce:" + guid;
        String token = CreateToken(message, apiSecret);
        System.Console.WriteLine("****************************************\n\nRequest\n-------\n" + message);
        System.Console.WriteLine("\nAccess Token:: " + token);

        string url = apiUrl + "/profiles/" + id;
        string output = string.Empty;

        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.Method = "DELETE";
        request.Headers["PaymentService-Date"] = datetime;
        request.Headers["PaymentService-Nonce"] = guid.ToString();
        request.Headers["Authorization"] = "Signature " + apiKey + ":" + token;

        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        using (Stream stream = response.GetResponseStream())
        using (StreamReader reader = new StreamReader(stream))
        {
            output = reader.ReadToEnd();
        }
        Console.WriteLine("\nResponse\n--------\n" + output);
    }


}
